src/codeeditor/editor.ts

```diff
@@ -0,0 +1,637 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+
+import {
+  IDisposable
+} from 'phosphor/lib/core/disposable';
+
+import {
+  find
+} from 'phosphor/lib/algorithm/searching';
+
+import {
+  Message
+} from 'phosphor/lib/core/messaging';
+
+import {
+  ISignal, clearSignalData, defineSignal
+} from 'phosphor/lib/core/signaling';
+
+import {
+  Widget
+} from 'phosphor/lib/ui/widget';
+
+import {
+  IChangedArgs
+} from '../common/interfaces';
+
+import {
+  ObservableVector
+} from '../common/observablevector';
+
+import {
+  JSONObject
+} from 'phosphor/lib/algorithm/json';
+
+
+/**
+ * A namespace for code editors.
+ */
+export
+namespace CodeEditor {
+  /**
+   * A zero-based position in the editor.
+   */
+  export
+  interface IPosition {
+    /**
+     * The cursor line number.
+     */
+    line: number;
+
+    /**
+     * The cursor column number.
+     */
+    column: number;
+  }
+
+  /**
+   * An interface describing editor state coordinates.
+   */
+  export
+  interface ICoords {
+    /**
+     * The left coordinate value.
+     */
+    readonly left: number;
+
+    /**
+     * The right coordinate value.
+     */
+    readonly right: number;
+
+    /**
+     * The top coordinate value.
+     */
+    readonly top: number;
+
+    /**
+     * The bottom coordinate value.
+     */
+    readonly bottom: number;
+  }
+
+  /**
+   * A text selection.
+   */
+  export
+  interface ITextSelection {
+    /**
+     * The index to the first character in the current selection.
+     *
+     * #### Notes
+     * If this position is greater than [end] then the selection is considered
+     * to be backward.
+     */
+    start: number;
+
+    /**
+     * The index to the last character in the current selection.
+     *
+     * #### Notes
+     * If this position is less than [start] then the selection is considered
+     * to be backward.
+     */
+    end: number;
+
+    /**
+     * The uuid of the text selection owner.
+     */
+    uuid: string;
+
+    /**
+     * A class name added to the text selection.
+     */
+    className?: string;
+  }
+
+  /**
+   * An editor model.
+   */
+  export
+  interface IModel extends IDisposable {
+    /**
+     * A signal emitted when the value changes.
+     */
+    valueChanged: ISignal<IModel, IChangedArgs<string>>;
+
+    /**
+     * A signal emitted when a property changes.
+     */
+    mimeTypeChanged: ISignal<IModel, IChangedArgs<string>>;
+
+    /**
+     * The text stored in the model.
+     */
+    value: string;  // TODO: this should be an iobservablestring.
+
+    /**
+     * A mime type of the model.
+     */
+    mimeType: string;
```

Comment:
I think the model should *always* have a mimetype, since that is what we are standardizing on moving forward (instead of mode), and should be `text/plain` if not set.